textbook/2_grouping.qmd

```diff
@@ -0,0 +1,994 @@
+
+:::
+
+Let's examine how jittering works.
+
+::: {.panel-tabset}
+
+# Altair
+
+In Altair, we create a random calculated variable called `jitter` by appending the `.transform_calculate()` method to our chart. We use our newly created `jitter` variable, encoding it as the `yOffset` of each data point using `yOffset='jitter:Q'`. This shifts the `yOffset` of each point by a random value. 
+
+The `:Q` simply tells Altair that `jitter` is a continuous quantitative variable.
+
+```{python}
+alt.Chart(income_df).mark_point().encode(
+    x='percentage_heavy_drinkers',
+    y='region',
+    color='region',
+    yOffset="jitter:Q"
+).transform_calculate(
+    jitter="random()"
+)
+```
+
+# ggplot
+
+In ggplot2, you can achieve the same effect with jittering by using the `position_jitter()` function. Jitter is directly applied to the marks through the position argument within `geom_point()`.
+
+```{r fig.width=8, fig.height=6}
+ggplot(income_df, aes(
+    x=percentage_heavy_drinkers, 
+    y=region, 
+    color=region
+)) + 
+geom_point(position = position_jitter(width = 0, height = 0.2))
+```
+
+:::
+
+
+::: {.callout-caution}
+Jittering can be confusing! It can easily be mistaken for a scatterplot with true variation in the underlying data.
+:::
+
+## Faceting
+
+Faceting takes the concept of grouping to the next level by generating multiple small charts, each dedicated to a subset of the data. It is particularly effective for highlighting distinct trends and patterns in each subset, as well as identifying extreme values or outliers within groups. And unlike grouping by position, faceting keeps both the x and y axis free for other variables.
+
+Let's check out the syntax:
+
+::: {.panel-tabset}
+
+# Altair
+
+In Altair, faceting by World Bank income group is straightforward. We simply append `.facet('income_group')` to our chart. Additionally, we can control the layout by setting the number of columns using the `columns` argument:
+
+```{python}
+alt.Chart(income_df).mark_point().encode(
+    x='percentage_heavy_drinkers',
+    y='life_expectancy_at_50',
+).facet(
+    'income_group',
+    columns=2
+)
+```
+
+# ggplot
+
+Facetting works via `facet_wrap`. By default it tries to create an even num of cols and rows, rather than putting all plot in one row like Altair. However, you can specify the desired number of columns with the `ncol` argument.
+
+
+```{r fig.width=8, fig.height=6}
+ggplot(income_df, aes(
+    x = percentage_heavy_drinkers, 
+    y = life_expectancy_at_50
+)) +
+geom_point() +
+facet_wrap(~ income_group, ncol=2)
+```
+
+:::
+
+Faceting makes it easy to see the overall shape of the distribution for each income group and how tightly clustered the values are. It also reveals outliers in the Upper Middle Income facet – we can easily spot two countries with a higher percentage of heavy drinkers.
+
+Another advantage is that we can apply additional layers of grouping within each facet. Now its your turn to try this!
```

Comment:
```suggestion
All the previous grouping strategies that we have discussed can be utilized within each facet as additional layers of grouping. This nested grouping can be powerful to stratify our data according to multiple categorical variables, but keep in mind that it can become confusing if we try to fit too much into a single visualization. Exactly how many levels of grouping you can use depends on how  well separated your data is and how many observations you have; more observations and less separation between groups means that you will have to be more careful and use less levels of grouping to keep the visualization easy to interpret.

Let's try out an example of how we can use faceting together with other grouping strategies.
```